src/halmos/sevm.py

```diff
@@ -1387,6 +1387,82 @@ def digest(self) -> bytes:
         return m.digest()
 
 
+class KeccakRegistry:
+    """A registry for tracking Keccak hash expressions and their corresponding values.
+
+    This class provides a dict-like interface for backward compatibility,
+    maintaining a mapping between hash values and their IDs (self._hashes).
+    This was the original functionality used for storing hash values and their corresponding IDs.
+
+    Additionally, this class provides an additional functionality for tracking
+    hash expressions and their generating values using an OffsetMap (self._exprs).
+    This allows for reverse lookups to find the original expression that
+    produced a given hash, even when the hash value is offset by some delta.
+    This is particularly useful for:
+    1. Local registry: Tracks hashes generated during execution
+    2. Precomputed registry: Contains known/common hash expressions
+
+    Example:
+        registry = KeccakRegistry()
+
+        # Storing hash values with ids (for backward compatibility)
+        registry[hash_value] = id_value
+        id_value = registry[hash_value]
+
+        # Tracking original hash expressions
+        registry.record(expr, hash_value)
+        original_expr = registry.reverse_lookup(hash_value)
+        # With offset
+        original_expr + delta = registry.reverse_lookup(hash_value + delta)
+    """
+
+    def __init__(self):
+        self._hashes: dict[BitVecRef, int] = {}  # hash -> id
+        self._exprs = OffsetMap()  # hash -> expr
+
+    def __getitem__(self, key: BitVecRef) -> int:
+        return self._hashes[key]
+
+    def __setitem__(self, key: BitVecRef, value: int) -> None:
+        self._hashes[key] = value
```

Comment:
would it make sense to add `assert (existing := self._hashes.get(key)) is None or existing == value`?

Reply:
yep i forgot about it